modules/Bio/EnsEMBL/Compara/Utils/CoreDBAdaptor.pm

```diff
@@ -109,6 +118,136 @@ sub pool_one_DBConnection {
+                          ' . ($genome_component ? $polyploid_extra_joins : '') . '
+                     WHERE at.code = "toplevel" AND cs.species_id = ?
+                          ' . ($genome_component ? $polyploid_extra_where: '');
+
+    my $sth = $dbc_copy->prepare($slice_sql, {'mysql_use_result' => 1});
+    $sth->bind_param(1, $species_dba->species_id(), SQL_INTEGER );
+    $sth->bind_param(2, $genome_component, SQL_VARCHAR) if $genome_component;
+    $sth->execute();
+    my ( $seq_region_id, $name, $length, $cs_id );
+    $sth->bind_columns( \( $seq_region_id, $name, $length, $cs_id ) );
+
+    my $slice_builder = sub {
+        if ($sth->fetch) {
+            my $cs = $csa->fetch_by_dbID($cs_id);
+            if(!$cs) {
+                throw("seq_region $name references non-existent coord_system $cs_id.");
+            }
+            return Bio::EnsEMBL::Slice->new_fast({
+                    'start'             => 1,
+                    'end'               => $length,
+                    'strand'            => 1,
+                    'seq_region_name'   => $name,
+                    'seq_region_length' => $length,
+                    'coord_system'      => $cs,
+                    'adaptor'           => $sa,
+                    # Not standard - Used by our own code below to record attributes
+                    'attributes'        => {
+                        'seq_region_id'     => $seq_region_id,
+                    },
+                });
+        }
+        $sth->finish;
+        $dbc_copy->disconnect_if_idle;
+        return;
+    };
+
+    my $slices_it = Bio::EnsEMBL::Utils::Iterator->new($slice_builder);
+    return $slices_it unless @$attributes;
+
+    my $batch_it = batch_iterator($slices_it, Bio::EnsEMBL::Compara::DBSQL::BaseAdaptor::ID_CHUNK_SIZE);
+
+    # Returns the same arrayref, but the slices will have the attributes loaded
+    my $attributes_fetcher = sub {
+        my $slices = shift;
+        my %slice_hash = map {$_->{'attributes'}->{'seq_region_id'} => $_} @$slices;
+        my $attrib_sql = 'SELECT seq_region_id, code, value
+                          FROM seq_region_attrib JOIN attrib_type USING (attrib_type_id)
+                          WHERE code IN (' . join(', ', map {"'$_'"} @$attributes) . ')
+                                AND seq_region_id IN (' . join(', ', keys %slice_hash) . ')';
+
+        $dbc->sql_helper->execute_no_return(
+            -SQL          => $attrib_sql,
+            -USE_HASHREFS => 1,
+            -CALLBACK     => sub {
+                my $row = shift;
+                my $slice = $slice_hash{$row->{'seq_region_id'}};
+                $slice->{'attributes'}->{$row->{'code'}} = $row->{'value'};
+            },
+        );
+        return $slices;
+    };
+    my $batch_with_attrib_it = $batch_it->map($attributes_fetcher);
+
+    return $return_batches ? $batch_with_attrib_it : flatten_iterator($batch_with_attrib_it);
```

Comment:
Batching is important internally when retrieving the attributes. With the batching on, it only needs to do n/500 queries instead of n (the roundtrip / overhead associated with each query is a killer)